### DISCUSSION TITLE
[Next 13/14] Bug with Pages Router and React library built with Parcel

### DISCUSSION THREAD

**Discussion by yyyyaaa** [vercel/next.js#57859] [Help]
### Summary

Hi all, I have a react library built with Parcel, which worked fine in next 12 but emits an error in next 13/14.
Reproduction steps:

Nextjs App:
https://github.com/yyyyaaa/reproduction-next-parcel-pages-bug

Library:
https://github.com/yyyyaaa/my-react-lib

- Clone both repo, run `npm install` in both repo
- in library folder, `run npm link && npm run build`
- in next app folder, run `npm link "my-react-lib"`, this linked the local library code. Then `npm run dev` , you'll see the error like below

<img width="1177" alt="Screenshot 2023-11-01 at 12 33 27" src="https://github.com/vercel/next.js/assets/16129111/ccfdbc70-23a2-4f56-b922-5bbaf4f97937">

I'm not sure if this is due to Parcel (something I need to tweak) or this is due to NextJS

### Additional information

Some googling lead me to these issues, which seem to be similar but no workable solutions brought up by the team or the community
https://github.com/vercel/next.js/issues/49355
https://github.com/tailwindlabs/headlessui/issues/2479

**Comment by icyJoseph** [comment]
Have you tried using `import {useRef, type ReactNode} from "react";` instead of `import * as React from "react";`?

**Reply by yyyyaaa** [reply]
hi @icyJoseph it did not help, why would that matter btw?

**Reply by icyJoseph** [reply]
It was just a guess tbh ~ Does changing the peerDependencies to react to something like `^18.0.0` help at all?

**Reply by icyJoseph** [reply]
Also, maybe look into using `bunchee` https://github.com/huozhi/bunchee, to discard Parcel grounded errors. (`swr` uses bunchee)

**Reply by yyyyaaa** [reply]
@icyJoseph nice, what do you mean Parcel grounded errors?

**Reply by icyJoseph** [reply]
As you were saying, maybe there's something that needs to be tweaked/fixed/or missing from Parcel's side. AFAIK SWR works with Next.js, so we can try to use the same tool, `bunchee`.

**Reply by yyyyaaa** [reply]
@icyJoseph bunchee uses rollup-plugin-dts, which is not maintained anymore :) I had a bug straight away (traced back to rollup-plugin-dts) after moving to bunchee lmao

**Reply by icyJoseph** [reply]
haha ~ library maintainers have to deal with this kind of thing, quite often :(

**Reply by yyyyaaa** [reply]
hey @icyJoseph I figured it out, somehow nextjs resolves 2 versions of React while in dev mode. I guess 1 for app router and 1 for pages router. Even if I'm just using page router it's still broken. This extra webpack config helped me solve it

```
const path = require("path");

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  /**
   *
   * @param {import('webpack').Configuration} config
   * @param {import('next/dist/server/config-shared').WebpackConfigContext} _context
   * @returns {import('webpack').Configuration}
   */
  webpack: (config, _context) => {
    const overridePath = path.resolve(__dirname, "../..", "node_modules/react");
    config.resolve.alias["react"] = overridePath;
    return config;
  },
};

module.exports = nextConfig;
```

**Reply by icyJoseph** [reply]
Aha maybe related? https://github.com/vercel/next.js/issues/56063#issuecomment-1740246934

**Reply by yyyyaaa** [reply]
I'm not sure lmao :D, spent too much time on this already. Another thing I found out is that Parcel is not the only one suffers from this bug, I tried a basic ViteJS/ Rollup library and it's the same. I hope this discussion will save time for somebody else
